File: src/LeetCode/Algorithms/Java/K_Closest_Points_to_Origin.java
package LeetCode.Algorithms.Java;

import java.util.ArrayList;
import java.util.Collections;

// https://leetcode.com/problems/k-closest-points-to-origin/
public class K_Closest_Points_to_Origin {
	public static void main(String[] args) {
		/*int[][] points = {{1, 3},
						{-2, 2}};
		int K = 1;*/
		/*int[][] points = {{3, 3},
						{5, -1},
						{-2, 4}};
		int K = 2;*/
		int[][] points = {{0, 1},
						{1, 0}};
		int K = 2;
		int[][] result = kClosest(points, K);
		
		for (int[] arr : result) {
			System.out.println(arr[0] + " " + arr[1]);
		}
	}
	
	public static int[][] kClosest(int[][] points, int K) {
		ArrayList<Integer> list = new ArrayList<>();
		int[][] result = new int[K][2];
		
		for (int[] pointArr : points) {
			list.add((pointArr[0] * pointArr[0]) + (pointArr[1] * pointArr[1]));
		}
		
		Collections.sort(list);
		
		int maxDistance = list.get(K - 1);
		int index = 0;
		
		for (int[] pointArr : points) {
			if ((pointArr[0] * pointArr[0]) + (pointArr[1] * pointArr[1]) <= maxDistance) {
				result[index++] = pointArr;
			}
		}
		
		return result;
	}
}

File: src/LeetCode/Algorithms/JavaScript/K_Closest_Points_to_Origin.js
/**
 * https://leetcode.com/problems/k-closest-points-to-origin/
 * @param {number[][]} points
 * @param {number} K
 * @return {number[][]}
 */
var kClosest = function(points, K) {
	const arr = new Array();
	
	points.map(function (point) {
		const distance = (point[0] * point[0]) + (point[1] * point[1]);
		
		arr.push(distance);
	});
	
	arr.sort((a, b) => {
		return a - b;
	});
	
	const max = arr.find(((value, index) => {
		if (index === K - 1) {
			return value;
		}
	}));
	
	const result = new Array();
	
	for (var i = 0; i < points.length && K > 0; i++) {
		const distance = (points[i][0] * points[i][0]) + (points[i][1] * points[i][1]);
		
		if (distance <= max) {
			result.push([points[i][0], points[i][1]]);
			K--;
		}
	}
	
	return result;
};

const points = [[1, 3], [-2, 2]];
const K = 1;

console.log(kClosest(points, K));